squeal-postgresql/src/Squeal/PostgreSQL/Expression/Time.hs

```diff
@@ -126,6 +130,46 @@ makeTimestamptz ::
    , null 'PGint4, null 'PGint4, null 'PGfloat8 ] ---> null 'PGtimestamptz
 makeTimestamptz = unsafeFunctionN "make_timestamptz"
 
+{-|
+Truncate a timestamp with the specified precision
+
+>>> printSQL $ dateTrunc ("day" *: makeTimestamp (2010 :* 5 :* 6 :* 14 :* 45 *: 11.4))
+date_trunc((E'day' :: text), make_timestamp((2010 :: int4), (5 :: int4), (6 :: int4), (14 :: int4), (45 :: int4), (11.4 :: float8)))
+-}
+dateTrunc ::
+  '[ null 'PGtext, null 'PGtimestamp ] ---> null 'PGtimestamp
+dateTrunc = unsafeFunctionN "date_trunc"
+
+{-|
+Truncate a timestamptz with the specified precision
+
+>>> printSQL $ dateTrunctz ("day" *: makeTimestamptz (2010 :* 5 :* 6 :* 14 :* 45 *: 11.4))
+date_trunc((E'day' :: text), make_timestamptz((2010 :: int4), (5 :: int4), (6 :: int4), (14 :: int4), (45 :: int4), (11.4 :: float8)))
+-}
+dateTrunctz ::
+  '[ null 'PGtext, null 'PGtimestamptz ] ---> null 'PGtimestamptz
+dateTrunctz = unsafeFunctionN "date_trunc"
+
+{-|
+Convert a timestamptz to a different timezone using an interval offset
+
+>>>  printSQL $ (makeTimestamptz (2015 :* 6 :* 9 :* 3 :* 45 *: 11.4)) `atTimeZonetz` (interval_ (-1) Hours )
+(make_timestamptz((2015 :: int4), (6 :: int4), (9 :: int4), (3 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE (INTERVAL '-1.000 hours'))
+-}
+atTimeZonetz ::
+  Operator (null 'PGtimestamptz) (null 'PGinterval) (null 'PGtimestamp)
+atTimeZonetz = unsafeBinaryOp "AT TIME ZONE"
+
+{-|
+Convert a timestamp to a different timezone using an interval offset
+
+>>> printSQL $ (makeTimestamp (2009 :* 7 :* 22 :* 19 :* 45 *: 11.4)) `atTimeZone` (interval_ 8 Hours)
+(make_timestamp((2009 :: int4), (7 :: int4), (22 :: int4), (19 :: int4), (45 :: int4), (11.4 :: float8)) AT TIME ZONE (INTERVAL '8.000 hours'))
+-}
+atTimeZone ::
+  Operator (null 'PGtimestamp) (null 'PGinterval) (null 'PGtimestamptz)
+atTimeZone = unsafeBinaryOp "AT TIME ZONE"
```

Comment:
I wonder if this would be better or worse using a closed type family similar to `PGSum`, so we could unify with `atTimeZonetz` and also support `PGtimetz`?

```
{-
* Treat given time stamp without time zone as located in the specified time zone
* Convert given time stamp with time zone to the new time zone, with no time zone designation
* Convert given time with time zone to the new time zone
-}
atTimeZone ::
  Operator (null 'PGtimestamp) (null 'PGinterval) (null 'PGtimestamptz)
atTimeZone = unsafeBinaryOp "AT TIME ZONE"

type family PGAtTimeZone ty where
  PGSum 'PGtimestamptz = 'PGtimestamp
  PGSum 'PGtimestamp = 'PGtimestamptz
  PGSum 'PGtimetz = 'PGtimetz
  PGSum pg = TypeError
    ( 'Text "Squeal type error: AT TIME ZONE cannot be applied to "
      ':<>: 'ShowType pg )
```